### PR TITLE
Packaging node-monitoring

### PR DESCRIPTION
**Pull request by TimothyAsirJeyasing**
Add spec file, makefile for package building
Add provision to auto bumpup version

tendrl-bug-id: Tendrl/performance-monitoring#29
Signed-off-by: Timothy Asir J <tjeyasin@redhat.com>

### CODE REVIEW

**Comment by TimothyAsirJeyasing** [comment]
@ktdreyer, @r0h4n , @anmolbabu , @nthomas-redhat , @shtripat , Please review

**Comment by ktdreyer** [comment]
I'm so confused. What is the difference between "node-monitoring" and "performance-monitoring"? Are there two projects embedded within this single GitHub repository?

**Comment by anmolbabu** [comment]
@ktdreyer 
> What is the difference between "node-monitoring" and "performance-monitoring"?
node-monitoring:
It is the package that needs to be installed on tendrl managed nodes. This package brings in the following:
* collectd configuration generation python script which will be installed on node as a command in /usr/bin/
* flow to generate collectd configuration using above mentioned command and other atoms involved in the flow which
* collectd custom plugin to handle(push alert to node_agent exposed socket) collectd detected threshold breaches.
* collectd templates for generating configurations.
* Pick and execute monitoring jobs targeted for node on which it resides. 
 
Performance-monitoring:
* This is responsible for aggregating collectd collected atomic statistics to composite stats for tendrl dashboards
* Configure collectd on tendrl managed nodes via node-monitoring
* Expose apis for 
  ** threshold configurations @ cluster level on end nodes
  ** access collectd collected stats from time series db.
  ** expose monitoring data points for tendrl ui

Yes both node-monitoring and performance-monitoring are part of same github repo.
The reason for this approach was to reduce having to maintain a large number of repos.
@brainfunked can you please comment?

**Comment by ktdreyer** [comment]
> Yes both node-monitoring and performance-monitoring are part of same github repo.
> The reason for this approach was to reduce having to maintain a large number of repos.

I recommend splitting these out into separate GitHub repositories. The `make snapshot` parts of the CI system determine the version to use based on Git tags. In this case, the Git tags will be applied to both projects :(

**Comment by anmolbabu** [comment]
@ktdreyer I think doing this might warrant bringing in some unexpected changes and due to brevity of time I feel its better to go with this approach for now and later split it into 2 github repos. @Tendrl/tendrl-core  what do you suggest?

**Comment by ktdreyer** [comment]
What are the exact challenges you face with splitting this up now? Maybe we can write some tooling to address the challenges you're facing, so we still keep your velocity and schedule? (Or maybe more people could have rights to create repos under the Tendrl org?)

If these two pieces really are separate packages designed to be installed separately, it makes the versioning mechanics here more difficult going forward.

**Comment by ktdreyer** [comment]
As a practical example, GitHub will send webhooks for this entire repository any time either `node-monitoring` or `performance-monitoring` change, so both packages will be rebuilt, even though they don't share any code. We won't be able to use Copr's built-in Mock SCM support with this configuration, like we can with the other Tendrl repos. The Jenkins job will have to diverge between this repository and every other repository. You won't be able to specify "pip install" `node-monitoring` directly from Git URLs.

If the packages were in the same namespace, I could see having one be a sub-package of another (with a single .spec file between them). But it doesn't appear that's the case here?

File: node_monitoring/tendrl-node-monitoring.spec [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+%{__python} setup.py build
+
+# remove the sphinx-build leftovers
+rm -rf html/.{doctrees,buildinfo}
+
+%install
+%{__python} setup.py install --single-version-externally-managed -O1 --root=$RPM_BUILD_ROOT --record=INSTALLED_FILES
+install -m  0755  --directory $RPM_BUILD_ROOT%{_sysconfdir}/collectd_template
+install -m  0755  --directory $RPM_BUILD_ROOT%{_usr}/lib64/collectd
+install -m  0755  --directory $RPM_BUILD_ROOT%{_usr}/bin/config_manager
```
Comment (anmolbabu) [inline node_monitoring/tendrl-node-monitoring.spec]: This should not be a directory.

I mean tendrl/node_monitoring/commands/config_manager.py should be copied as /usr/bin/config_manager

File: node_monitoring/tendrl-node-monitoring.spec [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+install -m  0755  --directory $RPM_BUILD_ROOT%{_var}/log/tendrl/node-monitoring
+install -Dm 0655 etc/tendrl/node-monitoring/node-monitoring.conf.yaml.sample $RPM_BUILD_ROOT%{_sysconfdir}/tendrl/node-monitoring/node-monitoring.conf.yaml
+install -Dm 0655 etc/tendrl/node-monitoring/logging.yaml.timedrotation.sample $RPM_BUILD_ROOT%{_sysconfdir}/tendrl/node-monitoring/node-monitoring_logging.yaml
+install -Dm 0655 tendrl/node_monitoring/commands/config_manager.py $RPM_BUILD_ROOT/usr/bin/config_manager
```
Comment (anmolbabu) [inline node_monitoring/tendrl-node-monitoring.spec]: what this does is it copies config_manager.py to /usr/bin/config_manager/ directory

File: node_monitoring/requirements.txt [hunk 3]
```diff
@@ -2,9 +2,6 @@
 # of appearance. Changing the order has an impact on the overall integration
 # process, which may cause wedges in the gate later.
 
-collectd
```
Comment (shtripat) [inline node_monitoring/requirements.txt]: From where do we get these removed dependencies now?

Comment (TimothyAsirJeyasing) [inline node_monitoring/requirements.txt]: Anmol sent a patch on top of this to fix requirements.txt errors. I will not touch this requirements.txt here.

File: node_monitoring/tendrl-node-monitoring.spec [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+License: LGPLv2+
+URL: https://github.com/Tendrl/performance-monitoring/tree/master/node_monitoring
+
+BuildRequires: pytest
+BuildRequires: systemd
+BuildRequires: python-mock
+
+Requires: ansible
+Requires: collectd
+Requires: collectd-ping
```
Comment (anmolbabu) [inline node_monitoring/tendrl-node-monitoring.spec]: Please remove this

Comment (TimothyAsir) [inline node_monitoring/tendrl-node-monitoring.spec]: Sure

File: node_monitoring/requirements.txt [hunk 5]
```diff
@@ -2,6 +2,9 @@
 # of appearance. Changing the order has an impact on the overall integration
 # process, which may cause wedges in the gate later.
 
-namespaces
-ansible
+collectd
+collectd-ping
```
Comment (TimothyAsirJeyasing) [inline node_monitoring/requirements.txt]: actually i have reverted the changes for requirements.txt, not doing anything or not touching anything in requirements.txt becuase anmol has the change in his patch

Comment (anmolbabu) [inline node_monitoring/requirements.txt]: @nthomas-redhat Tim has removed this file from his latest commit

File: node_monitoring/tendrl-node-monitoring.spec [hunk 6]
```diff
@@ -11,9 +11,6 @@ BuildRequires: pytest
 BuildRequires: systemd
 BuildRequires: python-mock
 
-Requires: ansible
-Requires: collectd
```
Comment (anmolbabu) [inline node_monitoring/tendrl-node-monitoring.spec]: @TimothyAsir Why did you remove ansible and collectd?

Comment (TimothyAsirJeyasing) [inline node_monitoring/tendrl-node-monitoring.spec]: @anmolbabu  you only asked me to remove this in the previous review, Please check your previous review

Comment (anmolbabu) [inline node_monitoring/tendrl-node-monitoring.spec]: @TimothyAsir My comment was to only remove collectd-ping

Comment (TimothyAsirJeyasing) [inline node_monitoring/tendrl-node-monitoring.spec]: :) Ok fine, i will update this patch

File: node_monitoring/requirements.txt (hunk 5 above)
Comment (nthomas-redhat) [inline node_monitoring/requirements.txt]: ok

**Comment by nthomas-redhat** [comment]
@TimothyAsirJeyasing , Can you raise a github issue to track the concern raised by @ktdreyer . Put the details over there, let us discuss and close on that issue. please link that issue here so that I can approve and merge the PR

**Comment by TimothyAsirJeyasing** [comment]
@nthomas-redhat , Yes @ktdreyer points are very much valid. I have raised an issue in https://github.com/Tendrl/performance-monitoring/issues/47

File: node_monitoring/tox.ini [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+[testenv:pep8]
+commands = flake8 {posargs}
+
+[testenv:venv]
+commands = {posargs}
+
+[testenv:cover]
+passenv = TRAVIS TRAVIS_JOB_ID TRAVIS_BRANCH
+#commands =
+#    py.test --cov-report= --cov=tendrl performance-monitoring/node-monitoring/tendrl/node-monitoring/tests
```
Comment (r0h4n) [inline node_monitoring/tox.ini]: why is this commented out? This basically means we wont run unit tests on new PRs, which is not acceptable

Comment (TimothyAsirJeyasing) [inline node_monitoring/tox.ini]: I Agree, Currently there is no 'tests' directory available in the path
node-monitoring/tendrl/node-monitoring/tests

@anmolbabu any reason why the tests folder is absent.
I could see a tests directory in tendrl/tests. Can we point to this path.

**Comment by TimothyAsirJeyasing** [comment]
@r0h4n , @nthomas-redhat , @shtripat , @anmolbabu , Please review

File: node_monitoring/tox.ini [hunk 8]
```diff
@@ -24,9 +24,9 @@ commands = {posargs}
 
 [testenv:cover]
 passenv = TRAVIS TRAVIS_JOB_ID TRAVIS_BRANCH
-#commands =
-#    py.test --cov-report= --cov=tendrl performance-monitoring/node-monitoring/tendrl/node-monitoring/tests
-#    coveralls --verbose
+commands =
+    py.test --cov-report=  --cov=tendrl tendrl/tests
```
Comment (nthomas-redhat) [inline node_monitoring/tox.ini]: does 'tendrl/tests' folder exists?
Did you try building rpms, after this change?

Comment (TimothyAsirJeyasing) [inline node_monitoring/tox.ini]: @nthomas-redhat yes its exists https://github.com/Tendrl/performance-monitoring/tree/master/node_monitoring/tendrl/tests
and i could able to build the rpm

Comment (nthomas-redhat) [inline node_monitoring/tox.ini]: ok
